### ISSUE TITLE
Cross-compiling facette 0.4

### ISSUE THREAD

**Issue by clemensg** [facette/facette#317]
Hi,

I am currently using the facette master branch and cross-compile facette for ARMv7. So far so good, works great on our embedded board, reading RRD data stored by collectd! :+1: 

For 0.4-alpha however, I noticed that you switched to gb, which does not support cross compiling Go code, which uses CGO to link to C libraries like librrd. The developer decided against adding it to gb. (https://github.com/constabulary/gb/issues/532#issuecomment-246519616)

What do you think about either switching back to go build instead of gb, or replacing the C library dependencies with pure Go code? Or is there another option?

Thanks,
Clemens

**Comment by vbatoufflet** [comment]
Hi @clemensg,

I did move to `gb` to ease Go libraries management and vendoring.

I'll take a look at what could be done to address this problem.

**Comment by vbatoufflet** [comment]
Is there any change you can provide us with some build commands you use to cross-compile Facette (for both `master` and `0.4-alpha`)?

**Comment by clemensg** [comment]
Hi @vbatoufflet , thanks.

On master, I have the following workflow:
```shell
export PATH=/opt/mycrosstoolchain/armv7a-linux-gnueabihf/gcc-6.3.0-glibc-2.25-binutils-2.28-kernel-4.10-sanitized/bin:$PATH
cd $GOPATH/src/github.com/facette/facette
make clean
PREFIX=/path/to/sysroot-target/usr CC=arm-v7a-linux-gnueabihf-gcc GOOS=linux GOARCH=arm GOARM=7 CGO_ENABLED=1 make install
```

Due to the `#cgo pkg-config: librrd` line in rrd_c.go from ziutek/rrd, go build uses `-I/path/to/sysroot-target/usr/include -L/path/to/sysroot-target/usr/lib -lrrd -lz ...` when compiling the librrd C bindings.

On 0.4-alpha, I tried with the same workflow, but it does not work with gb. I think it linked to the host system libraries instead.

Thanks for looking into it!

**Comment by clemensg** [comment]
@vbatoufflet Were you able to cross compile with my instructions? If you have a Raspberry Pi or something similar, you could use one of the prebuilt toolchains (e.g. https://github.com/raspberrypi/tools/tree/master/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf) and cross-compile for it.

Replacing all cgo libraries with go-only libs is too much work imho (it is not only librrd, but also the SQLite bindings and more).
It's probably easier to switch back to go build. What do you think about that? Or maybe it's possible to support both building with gb and go build?

**Comment by clemensg** [comment]
👍 a6e71f1 I see you were working on this. Thank you! I'll try if cross-compiling 0.4-alpha works now.

**Comment by vbatoufflet** [comment]
Hi @clemensg,

I didn't had the time to test it, but I indeed removed any usage of GB in the `Makefile`. Let me know if it fixes the issue you encountered.

**Comment by clemensg** [comment]
Awesome, it works! Thank you. 👍  🎉 

Just one question: I could not find the `url_prefix` option in 0.4-alpha. (I can't serve it from the base URL because we already have an API endpoint at `/api`, so I have to set `url_prefix` to something like `/facette`)

**Comment by vbatoufflet** [comment]
Indeed, it's not reimplemented yet in this branch. Sorry.

I just wrote a note, not to forget. I'll do that as quickly as possible.

**Comment by clemensg** [comment]
Great, thanks!

**Comment by vbatoufflet** [comment]
Hi @clemensg,

Just for the record, a `root_path` setting has been implemented in this branch (see 285e9985243bae59d46265c6ee4f13dfb1593d8e).

**Comment by clemensg** [comment]
Hi, thanks! The root_path setting works!

However, I found two other problems with 0.4-alpha:
- After adding graphs, they are not visible in the Graphs list in the admin panel or on the main page. I can add it to a collection, but that collection is not visible either.
- When adding a rrd provider with an empty filter list, it crashes. I added the stderr output below.

```
Apr 28 13:41:59 facette[521]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x36dbf4]
Apr 28 13:41:59 facette[521]: goroutine 189 [running]:
Apr 28 13:41:59 facette[521]: facette/catalog.NewFilterChain(0x0, 0x568d8540)
Apr 28 13:41:59 facette[521]:         /home/clemens/go/src/github.com/facette/facette/build/facette-linux-arm/src/facette/catalog/filter.go:63 +0x13c
Apr 28 13:41:59 facette[521]: main.newProviderWorker(0x56b9dca0, 0x568d4120, 0x568d8420, 0x24, 0x76c9a8)
Apr 28 13:41:59 facette[521]:         /home/clemens/go/src/github.com/facette/facette/build/facette-linux-arm/src/cmd/facette/poller_provider.go:46 +0x1f8
Apr 28 13:41:59 facette[521]: main.(*pollerWorker).StartProvider(0x56b9dca0, 0x568d4120)
```

**Comment by vbatoufflet** [comment]
Crash has been fixed in 874d2fcb8434d1abe6e29b1eae7b211b3098b625.

Regarding the first problem, I can't manage to reproduce it. If you think of a way, please open a new issue with the details.
